### PR TITLE
Re-authentication capability to avoid reusing short-lived access tokens

### PR DESCRIPTION
**Pull request by ddl-alexpanin**
Domino's python wrapper has `BearerAuth` class which reads the token from file during instantiation, and then keeps reusing it for all the subsequent calls. This doesn't work well with short-lived access tokens provided by JWT credentials propagation, when the token expires. Since Domino 4.5.0, access tokens are short-lived (5 minutes) by default.

Adding a decorator function which will read access token before calling any public library method.

Testing by installing library in a Domino workspace:

```
ubuntu@run-613853f46db72615aac81bdc-hc6k4:/mnt$ python3
Python 3.8.10 (default, Jun  4 2021, 15:09:15)
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from domino import Domino
>>> domino = Domino('integration-test/quick-start', host='<host>')
INFO:domino.domino:Domino deployment <host> is running version 5.0.0
>>> import time
>>> time.sleep(360)
>>> domino.collaborators_get()
['integration-test']
>>>
```

After commenting out lines 64-65 in modifiied `domino.py` and installing library in a Domino workspace:

```
ubuntu@run-613853f46db72615aac81bdc-hc6k4:/mnt$ python3
Python 3.8.10 (default, Jun  4 2021, 15:09:15)
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from domino import Domino
>>> domino = Domino('integration-test/quick-start', host='<host>')
INFO:domino.domino:Domino deployment <host> is running version 5.0.0
>>> import time
>>> time.sleep(360)
>>> domino.collaborators_get()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/.local/lib/python3.8/site-packages/domino/domino.py", line 66, in wrapper
    return fn(self, *args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/domino/domino.py", line 576, in collaborators_get
    return self._get(url)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/domino/domino.py", line 880, in _get
    return self.request_manager.get(url).json()
  File "/home/ubuntu/.local/lib/python3.8/site-packages/domino/http_request_manager.py", line 21, in get
    return self._raise_for_status(self.request_session.get(url, auth=self.auth, **kwargs))
  File "/home/ubuntu/.local/lib/python3.8/site-packages/domino/http_request_manager.py", line 34, in _raise_for_status
    response.raise_for_status()
  File "/opt/conda/lib/python3.8/site-packages/requests/models.py", line 941, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: <host>/integration-test/quick-start/collaborators
>>>
```

### CODE REVIEW

**Comment by mpkouznetsov** [comment]
Yes, using a decorator here is an elegant way to add the reauth aspect to all methods. I am not sure the way it is written right now actually works, since your decorator has to have access to `self` of `Domino`. Maybe something along these lines: https://stackoverflow.com/questions/11731136/class-method-decorator-with-self-arguments

**Comment by ddl-alexpanin** [comment]
> Yes, using a decorator here is an elegant way to add the reauth aspect to all methods. I am not sure the way it is written right now actually works, since your decorator has to have access to `self` of `Domino`. Maybe something along these lines: https://stackoverflow.com/questions/11731136/class-method-decorator-with-self-arguments

Works well as decorator belongs to the same class.

File: domino/domino.py [hunk 1]
```diff
@@ -74,15 +87,17 @@ def authenticate(self, api_key=None, auth_token=None, domino_token_file=None):
         self.request_manager = _HttpRequestManager(get_auth_by_type(api_key=api_key,
                                                                     auth_token=auth_token,
                                                                     domino_token_file=domino_token_file))
-
+    @_authenticate
```
Comment (dknupp) [inline domino/domino.py:90]: Thanks Alex. A few questions occurred to me. Have you considered the following?

- What happens if a user prefers to use the API token to authenticate instead of a token file?
- What if the user is using the client from an external node (i.e., not in the deployment)?
- Do we really want to insert a call to authenticate() at every invocation of every method?

Comment (ddl-alexpanin) [reply domino/domino.py:90]: > Thanks Alex. A few questions occurred to me. Have you considered the following?
> 
> * What happens if a user prefers to use the API token to authenticate instead of a token file?

`domino_token_file` will be empty and the `_authenticate` decorator will do nothing except calling method it is decorating, so no changes to the current behavior.

> * What if the user is using the client from an external node (i.e., not in the deployment)?

I think the user should be responsible to maintain the `domino_token_file` in such case.

Here is the context for the changes in this PR: https://dominodatalab.atlassian.net/browse/DOM-31477

> * Do we really want to insert a call to authenticate() at every invocation of every method?

It's not 100% elegant, and we can avoid calling it let's say for a method requiring authentication calling another such method, but do you think it's really worth doing? Our "authentication" basically does only reading token from the file.

File: domino/domino.py [hunk 2]
```diff
@@ -53,6 +56,16 @@ def __init__(self, project, api_key=None, host=None, domino_token_file=None, aut
             self._logger.error(error_message)
             raise Exception(error_message)
 
+    def _authenticate(fn):
+        """
+        Decorator for automatic re-authentication
+        """
+        def wrapper(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (ddl-alexpanin) [reply domino/domino.py:63]: >> The decorator assumes only one type of authentication -- token file. This is the my concern.

Yes and this is by design. There's nothing to do with the API key or auth_token at present time. We can rename `_authenticate` to something else to clarify it is only applicable to the token file case.

[the remark this answers is not in the document]
Comment (dknupp) [reply domino/domino.py:64]: I ran the unit tests on the `master` branch, and all tests passed.
```
platform linux -- Python 3.6.9, pytest-6.2.5, py-1.10.0, pluggy-1.0.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /mnt/python-domino, configfile: pytest.ini
plugins: requests-mock-1.9.3
collected 11 items                                                                                                                                             

tests/test_basic_auth.py::test_no_auth_type_error PASSED
tests/test_basic_auth.py::test_invalid_token_file_error PASSED
tests/test_basic_auth.py::test_malformed_project_input_error PASSED
tests/test_basic_auth.py::test_object_creation_with_auth_token PASSED
tests/test_basic_auth.py::test_object_creation_with_api_key PASSED
tests/test_basic_auth.py::test_object_creation_with_token_file PASSED
tests/test_basic_auth.py::test_re_authentication_with_a_new_type PASSED
tests/test_basic_auth.py::test_auth_token_precedence PASSED
tests/test_basic_auth.py::test_token_file_precedence PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_api_key PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_token_file PASSED
```
And then again on the `reauthn` branch.
```
tests/test_basic_auth.py::test_no_auth_type_error PASSED
tests/test_basic_auth.py::test_invalid_token_file_error PASSED
tests/test_basic_auth.py::test_malformed_project_input_error PASSED
tests/test_basic_auth.py::test_object_creation_with_auth_token PASSED
tests/test_basic_auth.py::test_object_creation_with_api_key PASSED
tests/test_basic_auth.py::test_object_creation_with_token_file PASSED
tests/test_basic_auth.py::test_re_authentication_with_a_new_type PASSED
tests/test_basic_auth.py::test_auth_token_precedence FAILED   # <----- this now fails
tests/test_basic_auth.py::test_token_file_precedence PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_api_key PASSED
tests/test_basic_auth.py::test_auth_against_real_deployment_with_token_file PASSED
```
...and as suspected, at least one test now fails. (I ran the tests in a Domino workspace. If I installed on my laptop and ran from there, I suspect more than one test would fail.)

I think this is the issue. As the class is defined now, the authentication type is one of the initial setup parameters.
```python
class Domino:
    def __init__(self, project, api_key=None, host=None, domino_token_file=None, auth_token=None):
```
So if I instantiate a new python-domino client like this:
```python
domino = Domino(host="https://my.host.com", project="me/quick-start", api_key="my_secret_key")
```
...the user's expectation (rightly so) is that their choice to authenticate with an API key is part of the saved state of that object.

But the decorator obliterates that initial input parameter by always forcing only one kind of authentication -- token file only.

In summary, I guess my general feeling is this is a good start, but it feels to me like more of a quick fix with "happy path" testing to solve one specific issue, without holistically taking into consideration the whole class.

[the remark this answers is not in the document]
Comment (ddl-alexpanin) [reply domino/domino.py:64]: Well okay, if you're saying that precedence is real and not artificial thing (never heard of anyone really supplying more than one artifact for authentication purposes), then of course we should do what I mentioned before.

[the remark this answers is not in the document]
Comment (dknupp) [reply domino/domino.py:65]: I agree that more supplying more than on authentication method at one time would be weird. If you do so now, the order of precedence is 1) API key, 2) token file, 3) auth token string -- but in general, it should not be done.

However, deciding to use API token (for example) is a real thing. And in fact, all of the tests QE runs to validate python-domino use that method, and they are most often run from a test client that is not part of the deployment.

After your most recent commit, unit tests are now passing if run from within a workspace. My feeling is still that this is a large change (because it inserts a new method call each and every time the API is accessed), so I personally still need some time to consider the full implications.

**Comment by ddl-alexpanin** [comment]
Closing in favor of even simpler solution.
